### PR TITLE
Add Get Stop Request.

### PR DESCRIPTION
**Pull request by siarheishut**
Process stat request:

- Stop <stop_name>.

Tested with unit-tests and manually.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -9,8 +9,10 @@ int main() {
   BusManager bm(ReadInputRequests(std::cin));
 
   for (auto &request : ReadOutputRequests(std::cin)) {
-    ProcessRequest(bm, request, std::cout);
+    if (auto ptr_b = std::get_if<GetBusRequest>(&request))
+      ProcessBusRequest(bm, *ptr_b, std::cout);
+    else if (auto ptr_s = std::get_if<GetStopRequest>(&request))
+      ProcessStopRequest(bm, *ptr_s, std::cout);
```
Comment (foxfrikses) [inline src/main.cpp:15]: As for me, I believe it still makes sense to have `ProcessRequest` function, which would only have
```
if (auto ptr_b = std::get_if<GetBusRequest>(&request))
  ProcessBusRequest(bm, *ptr_b, std::cout);
else if (auto ptr_s = std::get_if<GetStopRequest>(&request))
  ProcessStopRequest(bm, *ptr_s, std::cout);
```

File: src/bus_manager.h [hunk 2]
```diff
@@ -36,6 +43,7 @@ class BusManager {
   };
   double ComputeDistance(const std::vector<std::string> &stops);
 
+  std::unordered_map<std::string, StopResponse> buses_for_stops;
   std::unordered_map<std::string, Coords> stop_coords_;
```
Comment (foxfrikses) [inline src/bus_manager.h:47]: How about creating a struct `StopInfo` and putting all the information, related to a particular stop, in one place?

File: tests/request_reader_test.cpp [hunk 3]
```diff
@@ -398,29 +443,37 @@ TEST(TestReadOutputRequests, TestRequests) {
...
           .want = {},
       },
       TestCase{
           .name = "Wrong format(request_count): following spaces",
-          .input = "at2 \n"
+          .input = "2 \n"
                    "Bus bus1\n"
                    "Bus bus2",
           .want = {},
       },
```
Comment (foxfrikses) [inline tests/request_reader_test.cpp:457]: Let's make it in a different way: let's make it work fine with following and leading spaces. But let's make sure that the tests fail when we have non-digital symbols.

Same for input requests.

**Review by foxfrikses — CHANGES_REQUESTED** [review]

File: src/main.cpp (hunk 1 above)
Comment (siarheishut) [reply src/main.cpp:15]: Done.

File: src/bus_manager.h (hunk 2 above)
Comment (siarheishut) [reply src/bus_manager.h:47]: Done.

File: tests/request_reader_test.cpp (hunk 3 above)
Comment (siarheishut) [reply tests/request_reader_test.cpp]: Done.

**Review by foxfrikses — CHANGES_REQUESTED** [review]

File: src/bus_manager.cpp [hunk 4]
```diff
@@ -56,11 +56,16 @@ BusManager::BusManager(std::vector<PostRequest> requests) {
 void BusManager::AddStop(const std::string &stop,
                          Coords coords) {
   constexpr double k = 3.1415926535 / 180;
-  stop_coords_.emplace(stop, Coords{coords.latitude * k, coords.longitude * k});
+  stop_info_[stop].coords = {coords.latitude * k, coords.longitude * k};
+
+  stop_info_.try_emplace(stop);
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/bus_manager.cpp]: Done.

File: src/bus_manager.cpp [hunk 5]
```diff
@@ -56,11 +56,16 @@ BusManager::BusManager(std::vector<PostRequest> requests) {
...
 }
 
 void BusManager::AddBus(const std::string &bus,
                         std::vector<std::string> stops) {
+  for (auto& stop : stops)
+    stop_info_[stop].buses.emplace(bus);
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/bus_manager.cpp]: Done.

File: src/bus_manager.h [hunk 6]
```diff
@@ -15,12 +16,19 @@ struct BusResponse {
...
+  std::set<std::string> buses;
+};
+
 class BusManager {
  public:
   explicit BusManager(std::vector<PostRequest> requests);
 
   std::optional<BusResponse> GetBusInfo(const std::string &bus) const;
 
+  std::optional<StopInfo> GetStopInfo(const std::string &stop) const;
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/bus_manager.h]: Done.

File: src/request_reader.cpp [hunk 7]
```diff
@@ -14,6 +14,7 @@
 namespace {
 std::optional<int> ToInt(std::string_view sv) {
   int ans;
+  sv = rm::TrimEnd(rm::TrimStart(sv));
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/request_reader.cpp]: Done.

File: src/request_reader.cpp [hunk 8]
```diff
@@ -133,4 +138,10 @@ std::optional<GetBusRequest> ParseGetBusRequest(std::string_view sv) {
   if (bus.empty()) return std::nullopt;
   return GetBusRequest{.bus = std::string(bus)};
 }
+
+std::optional<GetStopRequest> ParseGetStopRequest(std::string_view sv) {
+  std::string_view stop = ReadNextToken(sv, "");
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/request_reader.cpp:144]: Done.

File: tests/request_reader_test.cpp [hunk 9]
```diff
@@ -306,24 +349,40 @@ TEST(TestReadInputRequests, TestRequests) {
...
                    "Bus 2: stop6 - stop7 - stop8\n"
-                   "Bus 4: stop2 > stop4 > stop7 > stop12",
+                   "Bus 4: stop2 > stop4 > stop7 > stop2",
           .want = {},
       },
       TestCase{
-          .name = "Wrong format(request_count): leading spaces",
-          .input = " 3\n"
+          .name = "Request_counter: leading spaces",
+          .input = "  3\n"
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply tests/request_reader_test.cpp:365]: Done.

File: src/bus_manager.cpp [hunk 10]
```diff
@@ -78,11 +81,21 @@ std::optional<BusResponse> BusManager::GetBusInfo(const std::string &bus) const
...
 }
 
+std::optional<StopResponse> BusManager::GetStopInfo(const std::string &stop) const {
+  auto it = stop_info_.find(stop);
+  if (it == stop_info_.end())
+    return std::nullopt;
+
+  if (it->second.buses.empty()) return StopResponse{{}};
+
+  return StopResponse{it->second.buses};
```
Comment (foxfrikses) [inline src/bus_manager.cpp:91]: If `it->second.buses.empty()`, can't you just return `StopResponse{it->second.buses};` as well?

File: src/request_reader.cpp [hunk 11]
```diff
@@ -133,4 +138,11 @@ std::optional<GetBusRequest> ParseGetBusRequest(std::string_view sv) {
   if (bus.empty()) return std::nullopt;
   return GetBusRequest{.bus = std::string(bus)};
 }
+
+std::optional<GetStopRequest> ParseGetStopRequest(std::string_view sv) {
+  // If empty "" delim is provided, just trims the string.
```
Comment (foxfrikses) [inline src/request_reader.cpp:143]: Not in front of this line, but in place, where you declare `ReadNextToken`

File: tests/request_reader_test.cpp [hunk 12]
```diff
@@ -302,28 +345,52 @@ TEST(TestReadInputRequests, TestRequests) {
...
       TestCase{
-          .name = "Wrong format(request_count): leading spaces",
-          .input = " 3\n"
+          .name = "Wrong format(request_count): following non-digital symbols",
+          .input = "12test\n"
                    "Stop stop1: 18.407908, 23.355151\n"
                    "Bus 2: stop6 - stop7 - stop8\n"
-                   "Bus 4: stop2 > stop4 > stop7 > stop12",
+                   "Bus 4: stop2 > stop4 > stop7 > stop2",
           .want = {},
```
Comment (foxfrikses) [inline tests/request_reader_test.cpp:361]: And how about adding the same for output requests too?

**Review by foxfrikses — CHANGES_REQUESTED** [review]

**Pull request merged**